python/taichi/ui/scene.py

```diff
@@ -13,6 +13,27 @@
 normals_field_cache = {}
 
 
+class DisplayMode_:
+    Fill = _ti_core.DisplayMode.Fill if _ti_core.GGUI_AVAILABLE else None
```

Comment:
Your DisplayMode shouldn't be dependent on whether GGUI_AVAILABLE IMHO. The enum can be always available, whether it's usable is controlled by GGUI availability.

Reply:
Still need it because every type in GGUI should check it's availability. And previously, I submitted the code without checking `GGUI_AVAILABLE` failed on some platforms.
ref:
1. https://github.com/taichi-dev/taichi/blob/7a8a9283535fbe9a17ee46132acd2fa4e7555c1f/python/taichi/ui/scene.py#L75-L80
2. https://github.com/taichi-dev/taichi/blob/7a8a9283535fbe9a17ee46132acd2fa4e7555c1f/python/taichi/ui/ui.py#L29
3. https://github.com/taichi-dev/taichi/blob/7a8a9283535fbe9a17ee46132acd2fa4e7555c1f/python/taichi/ui/ui.py#L12-L24

Reply:
For the second case about ProjectionMode I think we'd better remove the GGUI_AVAILABLE check. 
Scattering this check around isn't the best practice and we only need to check it when we create a GGUI resource like scene or camera you mentioned above (which really depends on whether it's available), but these enums can be available no matter what. 
For example, in the code snippet you showed in the pr description, you should check GGUI availability when you create a window or scene, instead of when creating a displaymode enum.

Reply:
personally speaking only the checks in `taichi.ui.Window.__init__` is necessary because the entire GGUI is nonsense without it. And if you don't have a clear idea how to pass parameters , `string` is always a go-to solution.